### PR TITLE
ASoC: Intel: sdw_rt711_rt1308_rt715: add multi-cpu DAI support

### PR DESCRIPTION
**Pull request by RanderWang**
In original code, one BE DAI is generated for each sdw link with
only one CPU DAI. But on some platforms, multiple speakers on
different sdw links are designed to support one PCM stream. In
this case there are multiple CPU DAI for each sdw link in one
BE DAI.

This driver processes each sdw link info and creates one BE DAI.
It is not fit for above multi-CPU-DAI case. This patch introduce
another sdw dai link, which merges multi-codec on different sdw
link into one sdw dai so that it is easy to gerenate BE DAI with
enough information.

Signed-off-by: randerwang <rander.wang@linux.intel.com>

### CODE REVIEW

**Comment by RanderWang** [comment]
Tested on playback CML-U, TGL-i2s and TGL-sdw.  The dai link is setup correctly on CML-H, but playback can't work on this branch.
```
[   48.800119] rt711 sdw:0:25d:711:0: DPN_PortCtrl register write failed for port 3
[   48.800129] int-sdw int-sdw.0: Program transport params failed: -61
[   48.800133] int-sdw int-sdw.0: Program params failed: -61
[   48.800137] int-sdw int-sdw.0: ASoC: cpu DAI prepare error: -61
[   48.800141]  SDW0-Playback: ASoC: backend prepare failed -61
````

File: sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c [hunk 1]
```diff
@@ -61,6 +61,7 @@ enum {
 #define SOF_SSP_PORT(x)		(((x) & GENMASK(5, 0)) << 5)
 #define SOF_SSP_GET_PORT(quirk)	(((quirk) >> 5) & GENMASK(5, 0))
 #define SOF_RT715_DAI_ID_FIX		BIT(11)
+#define SOF_SDW_MERGE_SPK_DAI		BIT(12)
```
Comment (plbossart) [inline sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c:64]: SOF_SDW_SPK_MULTI_DAI ?

File: sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c [hunk 2]
```diff
@@ -70,6 +71,12 @@ struct mc_private {
 	struct snd_soc_jack sdw_headset;
 };
 
+struct sdw_dai_link {
+	int dai_id;
+	int num_adr;
+	u64 *adr;
+};
```
Comment (plbossart) [inline sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c:78]: why do you need this, you already have all the information in mach_params? This looks like a lot of changes only to group the devices by dai_id?
it might be simpler to add a new field in the ACPI tables to say which devices are logically grouped

File: sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c [hunk 3]
```diff
@@ -234,7 +241,9 @@ static const struct dmi_system_id sof_sdw_rt711_rt1308_rt715_quirk_table[] = {
 			DMI_MATCH(DMI_SYS_VENDOR, "Dell Inc"),
 			DMI_MATCH(DMI_PRODUCT_NAME, "XPS"),
 		},
-		.driver_data = (void *)(SOF_RT711_JD_SRC_JD2),
+		.driver_data = (void *)(SOF_RT711_JD_SRC_JD2 |
+					SOF_RT715_DAI_ID_FIX |
```
Comment (plbossart) [inline sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c:245]: why did you change this? the ID_FIX has nothing to do with this?

File: sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c [hunk 4]
```diff
@@ -473,7 +482,6 @@ static const struct snd_soc_ops sdw_ops = {
 	.shutdown = sdw_shutdown,
 };
 
-static void rt711_init(const struct snd_soc_acpi_link_adr *link,
```
Comment (plbossart) [inline sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c:476]: something's missing here, this will not compile

File: sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c [hunk 5]
```diff
@@ -770,21 +812,112 @@ static int create_sdw_codec_dai(struct device *dev,
...
+static int convert_sdw_link_to_sdw_dai(struct device *dev,
+				       struct snd_soc_acpi_mach *mach,
+				       struct sdw_dai_link **sdw_dai,
+				       int *sdw_dai_num)
+{
+	const struct snd_soc_acpi_link_adr *adr_link;
+	const struct snd_soc_acpi_link_adr *link;
+	struct sdw_dai_link *tmp_link;
+	int spk_num, link_num;
+	u64 adr_data[4];
```
Comment (plbossart) [inline sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c:855]: why 4?

File: sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c [hunk 6]
```diff
@@ -770,21 +812,112 @@ static int create_sdw_codec_dai(struct device *dev,
...
+			spk_num = 0;
+			j++;
+		}
+	}
+
+	*sdw_dai = tmp_link;
+	*sdw_dai_num = link_num;
+
+	return 0;
+}
```
Comment (plbossart) [inline sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c:921]: sorry @RanderWang there's not a single comment and this looks like spaghetti code. please explain what this is supposed to do, there are just too many variables changing here.

File: sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c [hunk 7]
```diff
@@ -770,21 +812,112 @@ static int create_sdw_codec_dai(struct device *dev,
...
 	return 0;
 }
 
+static inline bool is_spk(int id)
+{
+	if (id == 0x1308)
+		return true;
+
+	return false;
+}
```
Comment (plbossart) [inline sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c:844]: oh gosh this is terrible, sorry... We should really not encode part numbers this way. I think we really want to have information in ACPI tables telling which devices are part of the same group.

BTW this is how it was supported by our Windows friends

**Review by plbossart — CHANGES_REQUESTED** [review]
Sorry @RanderWang this is just too experimental and needs more work.

**Comment by RanderWang** [comment]
@plbossart    I dumped the DSDT table.  windows driver create aggregated path by checking intel-endpoint-group-id != empty.  Now the intel-endpoint-group-id is not correct in ACPI setting. And
I also did many experiments in driver to read this value. I can't get this property. I studied ACPI driver
code (too complex) and printed all properties in SDW device, but I didn't find this property at all. I may spend much more time to study and check ACPI tree.

````
                    Package (0x02)
                    {
                        "intel-endpoint-group-id", 
                        Buffer (0x10)
                        {
                            /* 0000 */  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,  // ........
                            /* 0008 */  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00   // ........
                        }
                    }, 

                    Package (0x02)
                    {
                        "intel-endpoint-group-mask", 
                        Zero
                    }, 

                    Package (0x02)
                    {
                        "intel-endpoint-group-position", 
                        Zero
                    }, 

                    Package (0x02)
                    {
                        "intel-endpoint-direction", 
                        Zero
                    }, 
````

File: sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c (hunk 5 above)
Comment (RanderWang) [reply sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c:855]: it is a tentative setting,  max 4 speaker codec

File: sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c (hunk 4 above)
Comment (RanderWang) [reply sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c:476]: sorry, it is caused by merge conflict. I update PR now

File: sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c (hunk 3 above)
Comment (RanderWang) [reply sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c:245]: in original case , one 1308 is in dai 2, another 1308 is in dai 3, and dmic in dai 4. With mult-cpu dai,  two 1308s are in dai 2, so we need to fix dmic in dai 4.

File: sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c (hunk 1 above)
Comment (RanderWang) [reply sound/soc/intel/boards/sdw_rt711_rt1308_rt715.c:64]: multi-cpu DAI in one BE DAI, not multi-BE DAI.

**Comment by plbossart** [comment]
> @plbossart I dumped the DSDT table. windows driver create aggregated path by checking intel-endpoint-group-id != empty. Now the intel-endpoint-group-id is not correct in ACPI setting. And
> I also did many experiments in driver to read this value. I can't get this property. I studied ACPI driver
> code (too complex) and printed all properties in SDW device, but I didn't find this property at all. I may spend much more time to study and check ACPI tree.
> 
> ```
>                     Package (0x02)
>                     {
>                         "intel-endpoint-group-id", 
>                         Buffer (0x10)
>                         {
>                             /* 0000 */  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,  // ........
>                             /* 0008 */  0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00   // ........
>                         }
>                     }, 
> 
>                     Package (0x02)
>                     {
>                         "intel-endpoint-group-mask", 
>                         Zero
>                     }, 
> 
>                     Package (0x02)
>                     {
>                         "intel-endpoint-group-position", 
>                         Zero
>                     }, 
> 
>                     Package (0x02)
>                     {
>                         "intel-endpoint-direction", 
>                         Zero
>                     }, 
> ```

@RanderWang This sort of ACPI information will not be used in future generations (can't discuss this further in public), that's why I went for a simpler hard-coded version without dependencies on _DSD stuff.

**Comment by plbossart** [comment]
@RanderWang I would like you to start from https://github.com/thesofproject/linux/pull/1821, and use the 'group' information to figure out which cpu-dais need to be part of the same dailink.
